src/util/options.rs

```diff
@@ -389,105 +399,199 @@ impl FromStr for AffinityKind {
     /// bound. Hence, it is considered to be a "variable size" nursery. By default, a Bounded
     /// nursery has a lower bound of 2 MB and an upper bound of 32 MB for 32-bit systems and 1 TB
     /// for 64-bit systems.
-    Bounded,
+    Bounded {
+        /// The lower bound of the nursery size in bytes. Default to [`DEFAULT_MIN_NURSERY`].
+        min: usize,
+        /// The upper bound of the nursery size in bytes. Default to [`DEFAULT_MAX_NURSERY`].
+        max: usize,
+    },
+    /// A bounded nursery that is porportional to the current heap size. By default, a proportional bounded
+    // nursery has a lower bound of 20% of the heap size, and has an upper bound of 100% of the heap size.
+    ProportionalBounded {
+        /// The lower bound of the nursery size as a proportion of the current heap size. Default to [`DEFAULT_PROPORTIONAL_MIN_NURSERY`].
+        min: f64,
+        /// The upper bound of the nursery size as a proportion of the current heap size. Default to [`DEFAULT_PROPORTIONAL_MAX_NURSERY`].
+        max: f64,
+    },
     /// A Fixed nursery has the same upper and lower bounds. The size controls both the upper and
     /// lower bounds. Note that this is considered less performant than a Bounded nursery since a
     /// Fixed nursery size can be too restrictive and cause more GCs.
-    Fixed,
-}
-
-#[derive(Copy, Clone, Debug)]
-/// An option that provides a min/max interface to MMTk and a Bounded/Fixed interface to the
-/// user/VM.
-pub struct NurserySize {
-    /// The nursery type
-    pub kind: NurseryKind,
-    /// Minimum nursery size (in bytes)
-    pub min: usize,
-    /// Maximum nursery size (in bytes)
-    max: Option<usize>,
+    Fixed(usize),
 }
 
 impl NurserySize {
-    /// Create a NurserySize with the given kind. The value argument specifies the nursery size
-    /// for a fixed nursery, or specifies the max size for a bounded nursery.
-    pub fn new(kind: NurseryKind, value: Option<usize>) -> Self {
-        match kind {
-            NurseryKind::Bounded => NurserySize {
-                kind,
-                min: DEFAULT_MIN_NURSERY,
-                max: value,
-            },
-            NurseryKind::Fixed => NurserySize {
-                kind,
-                min: value.unwrap(),
-                max: value,
-            },
-        }
+    /// Estimate the virtual memory needed for the nursery space. This is used during space creation.
+    pub fn estimate_virtual_memory_in_bytes(&self) -> usize {
+        let virtual_memory_bytes = match *self {
+            NurserySize::Bounded { min: _, max } => max,
+            // Just use the default max nursery size -- the nursery won't get larger than that.
+            NurserySize::ProportionalBounded { min: _, max: _ } => {
+                if !crate::util::heap::vm_layout::vm_layout().force_use_contiguous_spaces {
+                    DEFAULT_MAX_NURSERY_32
```

Comment:
If the space is confined (not using contiguous spaces), the return value will be `DEFAULT_MAX_NURSERY_32` which is 32MB.  But the `Options::nursery` can be set to a larger proportion and can be larger than this.  As @k-sareen commented in https://github.com/mmtk/mmtk-core/pull/1087#discussion_r1548945572, `GCTrigger::get_max_nursery_bytes` may return a larger size.

Reply:
Yeah. You are right. I have changed this. I added a check: if we do not have pressure about virtual memory, the virtual memory for nursery can be as large as the max nursery (even if the max nursery is 100% heap size). Otherwise, it just panics.

Let me know what you think. @wks @k-sareen 

I will rerun the benchmarks for this change.

Reply:
Panicking is pretty drastic, but if we don't have a nice way of saying "the rest of the heap" then maybe for the time being it is fine. We should file an issue though. Also is there any way to test for a 32-bit architecture? The JikesRVM binding doesn't support generational collectors and the MockVM doesn't stress test the GC much.

Reply:
I think it might be better to panic instead of running with an unexpected value silently -- we could change this if the panic turns out to be an issue. 

What do you think we should test? The OpenJDK uses compressed pointer which tests most of the implementation for 32bits layout. I also added two unit tests that check the virtual memory we set for nursery.

Reply:
Yes I mainly wanted to test the virtual memory and ensuring that full-heap GCs happen more promptly. I guess OpenJDK with compressed pointers and the unit tests are good enough.